bumble/profiles/hap.py

```diff
@@ -0,0 +1,556 @@
+
+        self.active_preset_index = index
+        await self.notify_active_preset(connection)
+        return True
+
+    async def _on_set_active_preset(
+        self, connection: Optional[device.Connection], value: bytes
+    ):
+        await self.set_active_preset(connection, value)
+
+    async def set_next_preset(
+        self, connection: Optional[device.Connection], is_reverse
+        ):
+        ''' Set the next preset as active or previous if is_reverse is true'''
+        assert connection
+
+        if self.active_preset_index == 0x00:
+            raise att.ATT_Error(ErrorCode.PRESET_OPERATION_NOT_POSSIBLE)
+
+        first_preset: Optional[PresetRecord] = None  # To loop to first preset
+        next_preset: Optional[PresetRecord] = None
+        for index, record in sorted(self.preset_records.items(), reverse=is_reverse):
+            if not record.is_available():
+                continue
+            if first_preset == None:
+                first_preset = record
+            if index <= self.active_preset_index:
+                continue
+            next_preset = record
+            break
+
+        if not first_preset:  # if there is no first, there will be no next either
+            raise att.ATT_Error(ErrorCode.PRESET_OPERATION_NOT_POSSIBLE)
+
+        if next_preset:
+            self.active_preset_index = next_preset.index
+        else:
+            self.active_preset_index = first_preset.index
+        await self.notify_active_preset(connection)
+
+    async def _on_set_next_preset(self, connection: Optional[device.Connection]):
+        await self.set_next_preset(connection, False)
+
+    async def _on_set_previous_preset(self, connection: Optional[device.Connection]):
+        await self.set_next_preset(connection, True)
+
+    async def _on_set_active_preset_synchronized_locally(
+        self, connection: Optional[device.Connection], value: bytes
+    ):
+        assert (
+            self.hearing_aid_features.preset_synchronization_support
+            == PresetSynchronizationSupport.PRESET_SYNCHRONIZATION_IS_SUPPORTED
+        )
+        await self.set_active_preset(connection, value)
+        # TODO (low priority) inform other server of the change
+
+    async def _on_set_next_preset_synchronized_locally(
+        self, connection: Optional[device.Connection]
+    ):
+        assert (
+            self.hearing_aid_features.preset_synchronization_support
+            == PresetSynchronizationSupport.PRESET_SYNCHRONIZATION_IS_SUPPORTED
+        )
+        await self.set_next_preset(connection, False)
+        # TODO (low priority) inform other server of the change
+
+    async def _on_set_previous_preset_synchronized_locally(
+        self, connection: Optional[device.Connection]
+    ):
+        assert (
+            self.hearing_aid_features.preset_synchronization_support
+            == PresetSynchronizationSupport.PRESET_SYNCHRONIZATION_IS_SUPPORTED
+        )
```

Comment:
Prefer to raise specific error instead of assertion (unless it's a non-null assertion which is just for type checking).